src/query/api/v1/handler/namespace/update.go

```diff
@@ -0,0 +1,221 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"path"
+	"reflect"
+
+	clusterclient "github.com/m3db/m3/src/cluster/client"
+	"github.com/m3db/m3/src/cluster/kv"
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus/handleroptions"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3/src/query/util/logging"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"go.uber.org/zap"
+)
+
+var (
+	// M3DBUpdateURL is the url for the M3DB namespace update handler.
+	M3DBUpdateURL = path.Join(handler.RoutePrefixV1, M3DBServiceNamespacePathName)
+
+	// UpdateHTTPMethod is the HTTP method used with this resource.
+	UpdateHTTPMethod = http.MethodPut
+
+	fieldNameRetentionOptions = "RetentionOptions"
+	fieldNameRetetionPeriod   = "RetentionPeriodNanos"
+
+	errEmptyNamespaceName      = errors.New("must specify namespace name")
+	errEmptyNamespaceOptions   = errors.New("update options cannot be empty")
+	errEmptyRetentionOptions   = errors.New("retention options must be set")
+	errNamespaceFieldImmutable = errors.New("namespace option field is immutable")
+)
+
+// UpdateHandler is the handler for namespace updates.
+type UpdateHandler Handler
+
+// NewUpdateHandler returns a new instance of UpdateHandler.
+func NewUpdateHandler(
+	client clusterclient.Client,
+	instrumentOpts instrument.Options,
+) *UpdateHandler {
+	return &UpdateHandler{
+		client:         client,
+		instrumentOpts: instrumentOpts,
+	}
+}
+
+func (h *UpdateHandler) ServeHTTP(
+	svc handleroptions.ServiceNameAndDefaults,
+	w http.ResponseWriter,
+	r *http.Request,
+) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx, h.instrumentOpts)
+
+	md, rErr := h.parseRequest(r)
+	if rErr != nil {
+		logger.Error("unable to parse request", zap.Error(rErr))
+		xhttp.Error(w, rErr.Inner(), rErr.Code())
+		return
+	}
+
+	opts := handleroptions.NewServiceOptions(svc, r.Header, nil)
+	nsRegistry, err := h.Update(md, opts)
+	if err != nil {
+		logger.Error("unable to update namespace", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
```

Comment:
Should this be internal server error perhaps? 

I would've thought only parsing the request should result in a bad request error.